File: frontend/src/services/subscription.ts
import api from './api'

export interface Subscription {
  id: string
  user_id: string
  product_id: string
  league_id: string
  transaction_id?: string
  status: string
  started_at: string
  expires_at: string
  created_at: string
  product_name?: string
  league_name?: string
}

export interface SubscribeRequest {
  product_id: string
  league_id: string
  option_id?: string
}

export interface CheckAccessResponse {
  has_access: boolean
  subscription: Subscription | null
}

export const subscriptionService = {
  async subscribe(data: SubscribeRequest): Promise<Subscription> {
    const response = await api.post<Subscription>('/subscriptions', data)
    return response.data
  },

  async renew(id: string, optionId?: string): Promise<Subscription> {
    const response = await api.post<Subscription>(`/subscriptions/${id}/renew`, {
      option_id: optionId,
    })
    return response.data
  },

  async listMy(): Promise<{ subscriptions: Subscription[]; total: number }> {
    const response = await api.get<{ subscriptions: Subscription[]; total: number }>('/me/subscriptions')
    return response.data
  },

  async checkAccess(productId: string): Promise<CheckAccessResponse> {
    const response = await api.get<CheckAccessResponse>(`/products/${productId}/access`)
    return response.data
  },
}

export default subscriptionService

File: backend/internal/model/subscription.go
package model

import (
	"time"

	"github.com/google/uuid"
)

type Subscription struct {
	ID            uuid.UUID  `json:"id"`
	UserID        uuid.UUID  `json:"user_id"`
	ProductID     uuid.UUID  `json:"product_id"`
	LeagueID      uuid.UUID  `json:"league_id"`
	TransactionID *uuid.UUID `json:"transaction_id,omitempty"`
	Status        string     `json:"status"` // "active", "expired", "cancelled"
	StartedAt     time.Time  `json:"started_at"`
	ExpiresAt     time.Time  `json:"expires_at"`
	CreatedAt     time.Time  `json:"created_at"`

	// Joined fields
	ProductName   string `json:"product_name,omitempty"`
	LeagueName    string `json:"league_name,omitempty"`
	BuyerNickname string `json:"buyer_nickname,omitempty"`
	ProductPrice  *int64 `json:"product_price,omitempty"`
}

type SubscribeRequest struct {
	ProductID  uuid.UUID  `json:"product_id" validate:"required"`
	LeagueID   uuid.UUID  `json:"league_id" validate:"required"`
	OptionID   *uuid.UUID `json:"option_id,omitempty"`
	CouponCode *string    `json:"coupon_code,omitempty"`
}
